Allow ChipCertificateSet Initialization without Decode Buffer.

emargolis opened the pull request:
#### Problem
In the current implementation ChipCertificateSet always requires/allocates memory buffer even when it is not required.

#### Change overview
Updated implementation. 

#### Testing
-- Added new unit test
-- Updated some existing tests

Ticket: #8387

andy31415 commented on src/credentials/CHIPCert.cpp:
```diff
@@ -76,12 +76,15 @@ CHIP_ERROR ChipCertificateSet::Init(uint8_t maxCertsArraySize, uint16_t decodeBu
+        VerifyOrExit(mDecodeBuf != nullptr, err = CHIP_ERROR_NO_MEMORY);
+
+        mDecodeBufSize = decodeBufSize;
+    }
```
should there be an else to set mDecodeBuf to nullptr? or is that already guaranteed?

emargolis commented:
I've updated the initial fix. I removed the decodeBuf from Init() call as I observed it introduced a lot of confusion among developers on the purpose of this buffer and how it should be initialized. In the updated implementation the buffer is allocated on the stack for a brief duration only when TBS Hash is calculated.

Note: the original implementation was supposed to support decoding of any certificate sizes. This is not required for CHIP SDK because CHIP defines a fixed limit on a certificate size.

@bzbarsky-apple @andy31415 please review this PR again after the latest update and sorry for extra work.

tcarmelveilleux commented on src/credentials/CHIPCert.cpp:
```diff
@@ -187,11 +170,28 @@ CHIP_ERROR ChipCertificateSet::LoadCert(TLVReader & reader, BitFlags<CertDecodeF
+        // If requested to generate the TBSHash, verify that the decode buffer was initialized.
+        if (decodeFlags.Has(CertDecodeFlags::kGenerateTBSHash))
+        {
+            uint8_t decodeBuf[kMaxCHIPCertDecodeBufLength];
```
This is a lot of stack to add. This amount of stack may be fine, but it should be called-out.

@jmartinez-silabs Would it be possible to test that this PR doesn't overrun on your smallest part, in usage?

emargolis replied:
Another option would be to call:
    `decodeBuf = reinterpret_cast<uint8_t *>(chip::Platform::MemoryAlloc(kMaxCHIPCertDecodeBufLength));`
and then release.

tcarmelveilleux replied:
`ScopedMemoryBuffer` is what you want. It's an RAII allocator over MemoryAlloc.

emargolis replied:
Updated to use `ScopedMemoryBuffer`
